ocaml/typing/typetexp.ml

```diff
@@ -677,6 +677,20 @@ and transl_type_aux env policy mode styp =
       in
       List.iter2
         (fun (sty, cty) ty' ->
+           let get_reason layout = match Layout.get layout with
+             | Const Value -> Layout.(Value_creation (Imported_type_argument path))
+             (* CR layouts: Add more cases here when type params of imported types can
+                have non-value layout *)
+             | _ -> Layout.Imported
+           in
+           begin match Types.get_desc ty' with
+           | Tvar {layout; _} when Layout.has_imported_history layout ->
+             (* In case of a Tvar with imported layout history, we can improve
+                the layout reason using the in scope [path] to the parent type. *)
+             let reason = get_reason layout in
+             Types.set_var_layout ty' (Layout.update_reason layout reason)
```

Comment:
We've been worried about performance here. If you've done some experiments that suggest why performance is OK, summarize the results in a comment near here.